modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/tx/ScriptTransactionHandler.java

```diff
@@ -0,0 +1,319 @@
+import org.apache.ignite.tx.IgniteTransactions;
+import org.apache.ignite.tx.TransactionOptions;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * Starts an implicit transaction if there is no external transaction. Supports script transaction management using
+ * {@link SqlQueryType#TX_CONTROL} statements.
+ */
+public class ScriptTransactionHandler extends QueryTransactionHandler {
+    /** No-op transaction wrapper. */
+    private static final QueryTransactionWrapper NOOP_TX_WRAPPER = new NoopTransactionWrapper();
+
+    /** Wraps a transaction, which is managed by SQL engine via {@link SqlQueryType#TX_CONTROL} statements. */
+    private volatile @Nullable ManagedTransactionWrapper wrapper;
+
+    public ScriptTransactionHandler(IgniteTransactions transactions, @Nullable InternalTransaction externalTransaction) {
+        super(transactions, externalTransaction);
+    }
+
+    @Override
+    protected @Nullable InternalTransaction activeTransaction() {
+        return externalTransaction == null ? scriptTransaction() : externalTransaction;
+    }
+
+    /**
+     * Starts a transaction if there is no external transaction.
+     *
+     * @param parsedResult Parse result.
+     * @param cursorFut Cursor future for the current statement.
+     * @return Transaction wrapper.
+     */
+    public QueryTransactionWrapper startScriptTxIfNeeded(
+            ParsedResult parsedResult,
+            CompletableFuture<? extends AsyncCursor<?>> cursorFut
+    ) {
+        try {
+            SqlQueryType queryType = parsedResult.queryType();
+
+            if (queryType == SqlQueryType.TX_CONTROL) {
+                if (externalTransaction != null) {
+                    throw new ExternalTransactionNotSupportedException();
+                }
+
+                return handleTxControlStatement(parsedResult.parsedTree());
+            }
+
+            ManagedTransactionWrapper wrapper = this.wrapper;
+
+            if (wrapper == null) {
+                return startTxIfNeeded(queryType);
+            }
+
+            validateStatement(parsedResult.queryType(), wrapper.unwrap());
+
+            return wrapper.forStatement(queryType, cursorFut);
+        } catch (SqlException e) {
+            InternalTransaction scriptTx = scriptTransaction();
+
+            if (scriptTx != null) {
+                scriptTx.rollback();
+            }
+
+            throw e;
+        }
+    }
+
+    private @Nullable InternalTransaction scriptTransaction() {
+        ManagedTransactionWrapper hld = wrapper;
+
+        return hld != null ? hld.unwrap() : null;
+    }
+
+    private QueryTransactionWrapper handleTxControlStatement(SqlNode node) {
+        ManagedTransactionWrapper txWrapper = this.wrapper;
+
+        if (node instanceof IgniteSqlCommitTransaction) {
+            if (txWrapper == null) {
+                return NOOP_TX_WRAPPER;
+            }
+
+            this.wrapper = null;
+
+            return txWrapper.forCommit();
```

Comment:
looking at this place i expected test will pass, but it didn\`t, i\`m wrong ? i.e. i see that RO transaction already commited but we still have partially readed cursors.
```
    @Test
    void readMoreRowsThanCanBePrefetchedReadOnlyTx() {
        String query = "START TRANSACTION READ ONLY;"
                + "SELECT id FROM big;"
                + "SELECT 1;"
                + "COMMIT;";

        List<AsyncSqlCursor<List<Object>>> cursors = fetchAllCursors(runScript(query));
        assertThat(cursors, hasSize(4));
        verifyFinishedTxCount(0);

        BatchedResult<List<Object>> res = await(cursors.get(2).requestNextAsync(1));
        assertNotNull(res);
        verifyFinishedTxCount(0);

        res = await(cursors.get(2).requestNextAsync(100));
        assertNotNull(res);
        verifyFinishedTxCount(0);

        res = await(cursors.get(1).requestNextAsync(1));
        assertNotNull(res);
        verifyFinishedTxCount(0);

        cursors.forEach(AsyncSqlCursor::closeAsync);
    }
```

Reply:
We have decided that in the current implementation, only the RW transaction have limitation - that all cursors must be closed.  Because our current RO transactions are just a timestamp, and it looks like we don't need to block script execution with the RO transaction.

Reply:
ok, got it